src/scripts/antithesis.zig

```diff
@@ -0,0 +1,241 @@
+//! Deployment script for our systest (src/testing/systest).
+//!
+//! * Builds the Java client and the associated workload using Maven
+//! * Builds Docker images for the workload, replicas, and config
+//! * Optionally pushes the images to the Antithesis registry
+//!
+//! Currently there's no support for triggering tests with this script.
+
+const std = @import("std");
+const builtin = @import("builtin");
+const log = std.log;
+const assert = std.debug.assert;
+
+const stdx = @import("../stdx.zig");
+const flags = @import("../flags.zig");
+const fatal = flags.fatal;
+const Shell = @import("../shell.zig");
+
+pub const CLIArgs = struct {
+    tag: []const u8,
+    push: bool = false,
+};
+
+const Image = enum { config, workload, replica };
+
+pub fn main(shell: *Shell, gpa: std.mem.Allocator, cli_args: CLIArgs) !void {
+    assert(try shell.exec_status_ok("docker --version", .{}));
+
+    // Docker tag to build and push
+    const tag = std.mem.trim(u8, cli_args.tag, &std.ascii.whitespace);
+    assert(tag.len > 0);
+
+    try shell.zig("build -Drelease", .{});
+
+    // Build Java client library
+    {
+        try shell.pushd("./src/clients/java");
+        defer shell.popd();
+
+        try shell.exec("mvn clean install --batch-mode --quiet -Dmaven.test.skip", .{});
+    }
+
+    // Build workload
+    {
+        try shell.pushd("./src/testing/systest/workload");
+        defer shell.popd();
+
+        try shell.exec("mvn clean package --batch-mode --quiet", .{});
+    }
+
+    const images = comptime std.enums.values(Image);
+    inline for (images) |image| {
+        try build_image(shell, gpa, image, tag);
+        if (cli_args.push) {
+            try push_image(shell, image, tag);
+        }
+    }
+}
+
+fn build_image(
+    shell: *Shell,
+    gpa: std.mem.Allocator,
+    comptime image: Image,
+    tag: []const u8,
+) !void {
+    switch (image) {
+        .config => {
+            const image_dir = try shell.exec_stdout("mktemp -d", .{});
+
+            try shell.pushd(image_dir);
+            defer shell.popd();
+
+            // TODO(owickstrom): remove the need for .env file by rendering docker-compose.yaml
+            // with the correct tag?
+            const env_file = try std.fs.path.join(gpa, &.{ image_dir, ".env" });
+            defer gpa.free(env_file);
```

Comment:
You can use `shell.fmt` to avoid the need to free individual strings. I am very distrustfull of code that manages individual strings, it always leaks eventually.

Reply:
I am on the fence on whether we need `shell.join` as well -- the `/` separator _does_ work on windows, so I don't see problems with raw string formatting personally.